packages-next/testing/src/index.ts

```diff
@@ -0,0 +1,100 @@
+import path from 'path';
+import crypto from 'crypto';
+import fs from 'fs';
+import supertest, { Test } from 'supertest';
+import memoizeOne from 'memoize-one';
+import { initConfig, createSystem, createExpressServer } from '@keystone-next/keystone';
+import { pushPrismaSchemaToDatabase } from '@keystone-next/keystone/migrations';
+import {
+  getCommittedArtifacts,
+  writeCommittedArtifacts,
+  requirePrismaClient,
+  generateNodeModulesArtifacts,
+} from '@keystone-next/keystone/artifacts';
+import type { KeystoneConfig, KeystoneContext } from '@keystone-next/types';
+
+export type GraphQLRequest = (arg: {
+  query: string;
+  variables?: Record<string, any>;
+  operationName?: string;
+}) => Test;
+
+export type TestArgs = {
+  context: KeystoneContext;
+  config: KeystoneConfig;
+  graphQLRequest: GraphQLRequest;
+};
+
+export type TestEnv = {
+  connect: () => Promise<void>;
+  disconnect: () => Promise<void>;
+  testArgs: TestArgs;
+};
+
+const _hashPrismaSchema = memoizeOne(prismaSchema =>
+  crypto.createHash('md5').update(prismaSchema).digest('hex')
+);
+const _alreadyGeneratedProjects = new Set<string>();
+export async function setupTestEnv({
+  config: _config,
+}: {
+  config: KeystoneConfig;
+}): Promise<TestEnv> {
+  // Force the UI to always be disabled, re-spread the db option as we're going to
+  // modify it below.
+  const config = initConfig({ ..._config, db: { ..._config.db }, ui: { isDisabled: true } });
+  const { graphQLSchema, getKeystone } = createSystem(config);
+
+  const isolateRuns = config.testing?.isolateRuns || false;
+  let artifactPath = path.resolve(config.testing?.artifactPath || '.keystone/tests');
+
+  const artifacts = await getCommittedArtifacts(graphQLSchema, config);
+  const hash = _hashPrismaSchema(artifacts.prisma);
+
+  if (isolateRuns) {
+    if (config.db.provider === 'postgresql') {
+      config.db.url = `${config.db.url}?schema=keystone-test-${hash.toString()}`;
+    }
```

Comment:
What's the purpose of this? It doesn't allow running tests concurrently since two tests might have prisma schemas with the same hash and we drop the whole database anyway so this just seems like it'll fill up your local database with a bunch of stuff rather than just filling up a single postgres schema with the way we have it currently?

Reply:
This is largely an optimisation that might no longer be needed. When working with `knex` the step of recreating the schema itself was extremely slow, so recreating new schemas for each different test made local testing of the keystone project really painful when running locally. I haven't done any experiments since we moved to prisma, so this may no longer be such an issue. Would need to test it out.

Reply:
Oh, sorry, I thought this wasn't here before, nvm then.